keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,659 @@
+
+This proposal also aims to improve the initialization and reinitialization of resource managers, such as the CPU manager and memory manager, in response to changes in a node's CPU and memory configurations.
+
+## Motivation
+In a typical Kubernetes environment, the cluster resources may need to be altered due to following reasons:
+- Incorrect resource assignment during cluster creation.
+- Increased workload over time, leading to the need for additional resources in the cluster.
+
+To handle these scenarios, we can:
+- Horizontally scale up the cluster by adding compute nodes.
+- Vertically scale up the cluster by increasing node capacity. However, currently, the workaround for capturing node resizing in the cluster involves restarting the Kubelet.
+
+Node resource hot plugging will provide advantages in scenarios such as:
+- Handling resource demand with a limited set of nodes by increasing the capacity of existing nodes instead of creating new nodes.
+- Creating new nodes takes more time compared to increasing the capacity of existing nodes.
+
+### Goals
+
+* Dynamically scale up the node by hot plugging  resources and without restarting the kubelet.
+* Ability to reinitialize resource managers (CPU manager, memory manager) to adopt changes in node's resource.
+
+### Non-Goals
+
+* Dynamically adjust system reserved and kube reserved values.
+* Hot unplug of node resources.
+* Update the autoscaler to utilize resource hot plugging.
+
+
+## Proposal
+
+This KEP aims to support the node resource hot plugging by adding a polling mechanism in kubelet to fetch the machine-information from cAdvisor's cache which is already updated periodically, This information will be fetched periodically by kubelet, after which the node status updater is responsible for updating this information at node level in the cluster.
+
+Additionally, this KEP aims to improve the initialization and reinitialization of resource managers, such as the memory manager and CPU manager, so that they can adapt to change in node's configurations.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I must be able to increase the cluster resource capacity without adding a new node to the cluster.
+
+#### Story 2
+
+As a cluster admin, I must be able to increase the cluster resource capacity without need to restarting the kubelet.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+1. Node resource hot plugging is an opt-in feature, merging the
+   feature related changes won't impact existing workloads. Moreover, the feature
+   will be rolled out gradually, beginning with an alpha release for testing and
+   gathering feedback. This will be followed by beta and GA releases as the
+   feature matures and potential problems and improvements are addressed.
+2. Though the node resource is updated dynamically, the dynamic data is fetched from cAdvisor and its well integrated with kubelet.
```

Comment:
I saw discussions around pursuing CRI for obtaining resource information and not completely relying on Cadvisor. Can we capture in the KEP why we decided to go ahead with Cadvisor approach? I think we should still capture details related to CRI in Alternatives section.

Reply:
So when we initially brought up this feature request there were opinion on moving away from cAdvisor for various reasons like  cAdvisor data structure is not able to correctly represent a variety of modern CPU architectures. We are planning to separate KEP for that and work is in progress.

With either of them will just provide latest machine Info and we want to make kubelet consume and react to that decided to move with cAdvisor which is already integrated with kubelet.

Reply:
there is a KEP to remove cadvisor from the equation, but the node level metrics would be collected by the kubelet in that case. CRI stats (2371) doesn't have moving machine metrics to CRI in scope

Reply:
In this case I see some benefits in making the resource managers react to hotplug without changing the resource information, which is a related but different (and quite big) task.
I agree this wok should be mentioned somewhere in the KEP per @swatisehgal suggestion, though.